From 39cad715d8e0bfc6266ddf553fc1cc120181e227 Mon Sep 17 00:00:00 2001
From: Andrey Yepin <ayepin@google.com>
Date: Fri, 23 May 2025 14:05:41 -0700
Subject: [PATCH] Sanitize cross-profile intents.

Remove package or component information from payload intents (and their
selectors) for cross-profile sharing.

Bug: 407764858
Test: manual testing with test app
Flag: EXEMPT bugfix
Merged-In: Ie00f0a4e2fdb0cb7a3b2a148369d435d7da9b1bc
Change-Id: Ib153416d98325da300fe383b877374bc6b21ff90
---
 .../android/internal/app/ChooserActivity.java | 37 +++++++++++-
 ...hooserActivityCrossProfileIntentsTest.java | 59 +++++++++++++++++++
 2 files changed, 94 insertions(+), 2 deletions(-)
 create mode 100644 core/tests/coretests/src/com/android/internal/app/ChooserActivityCrossProfileIntentsTest.java

diff --git a/core/java/com/android/internal/app/ChooserActivity.java b/core/java/com/android/internal/app/ChooserActivity.java
index 4e4305aa73b1..2f54049547fd 100644
--- a/core/java/com/android/internal/app/ChooserActivity.java
+++ b/core/java/com/android/internal/app/ChooserActivity.java
@@ -857,16 +857,21 @@ public class ChooserActivity extends ResolverActivity implements
             List<ResolveInfo> rList,
             boolean filterLastUsed) {
         int selectedProfile = findSelectedProfile();
+        List<Intent> crossProfileIntents = sanitizePayloadIntents(mIntents);
         ChooserGridAdapter personalAdapter = createChooserGridAdapter(
                 /* context */ this,
-                /* payloadIntents */ mIntents,
+                /* payloadIntents */ selectedProfile == PROFILE_PERSONAL
+                        ? mIntents
+                        : crossProfileIntents,
                 selectedProfile == PROFILE_PERSONAL ? initialIntents : null,
                 rList,
                 filterLastUsed,
                 /* userHandle */ getPersonalProfileUserHandle());
         ChooserGridAdapter workAdapter = createChooserGridAdapter(
                 /* context */ this,
-                /* payloadIntents */ mIntents,
+                /* payloadIntents */ selectedProfile == PROFILE_WORK
+                        ? mIntents
+                        : crossProfileIntents,
                 selectedProfile == PROFILE_WORK ? initialIntents : null,
                 rList,
                 filterLastUsed,
@@ -4086,4 +4091,32 @@ public class ChooserActivity extends ResolverActivity implements
                 target.getComponentName(),
                 target.getIntentExtras());
     }
+
+    /**
+     * Returns a copy of provided intents with explicit targeting information is removed from each
+     * intent in the list, as well as from its selector {@link Intent#getSelector}. Specifically,
+     * the values that would be returned by {@link Intent#getPackage} and
+     * {@link Intent#getComponent} are cleared for both the main intent and its selector. This
+     * sanitization is performed because explicit intents could otherwise be used to bypass the
+     * device's cross-profile sharing policy settings.
+     */
+    @NonNull
+    @VisibleForTesting
+    public static List<Intent> sanitizePayloadIntents(@NonNull List<Intent> intents) {
+        return intents.stream().map((intent) -> {
+            if (intent == null) {
+                return null;
+            }
+            Intent sanitized = new Intent(intent);
+            sanitized.setPackage(null);
+            sanitized.setComponent(null);
+            if (sanitized.getSelector() != null) {
+                Intent selector = new Intent(sanitized.getSelector());
+                selector.setPackage(null);
+                selector.setComponent(null);
+                sanitized.setSelector(selector);
+            }
+            return sanitized;
+        }).collect(Collectors.toList());
+    }
 }
diff --git a/core/tests/coretests/src/com/android/internal/app/ChooserActivityCrossProfileIntentsTest.java b/core/tests/coretests/src/com/android/internal/app/ChooserActivityCrossProfileIntentsTest.java
new file mode 100644
index 000000000000..d022d48df9d4
--- /dev/null
+++ b/core/tests/coretests/src/com/android/internal/app/ChooserActivityCrossProfileIntentsTest.java
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.internal.app;
+
+import static android.content.Intent.ACTION_SEND;
+
+import static com.google.common.truth.Truth.assertThat;
+
+import android.content.ComponentName;
+import android.content.Intent;
+
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class ChooserActivityCrossProfileIntentsTest {
+    @Test
+    public void test_sanitizePayloadIntents() {
+        final Intent intentOne = new Intent(ACTION_SEND);
+        intentOne.setPackage("org.test.example");
+        final Intent intentTwo = new Intent(ACTION_SEND);
+        intentTwo.setComponent(ComponentName.unflattenFromString("org.test.example/.TestActivity"));
+        final Intent intentThree = new Intent(ACTION_SEND);
+        intentThree.setSelector(new Intent(intentOne));
+        final Intent intentFour = new Intent(ACTION_SEND);
+        intentFour.setSelector(new Intent(intentTwo));
+        ArrayList<Intent> intents = new ArrayList<>();
+        Collections.addAll(intents, intentOne, intentTwo, intentThree, intentFour);
+
+        List<Intent> sanitizedIntents = ChooserActivity.sanitizePayloadIntents(intents);
+
+        assertThat(sanitizedIntents).hasSize(intents.size());
+        for (Intent intent : sanitizedIntents) {
+            assertThat(intent.getPackage()).isNull();
+            assertThat(intent.getComponent()).isNull();
+            Intent selector = intent.getSelector();
+            if (selector != null) {
+                assertThat(selector.getPackage()).isNull();
+                assertThat(selector.getComponent()).isNull();
+            }
+        }
+    }
+}
-- 
2.50.1.351.gcf940e82a1-goog

